juju/juju
# Default ConfigSource implementation uses netlink

This patch adds a new implementation of `ConfigSource` in `core/network` that uses the `netlink` dependency also introduced here.

The default `ConfigSource` is switched to this new implementation for Linux builds, with others using the old implementation by way of build directives. This is necessary due to the netlink library not building on Windows. 

With this patch we can now use netlink's capability to surface the `IFA_F_SECONDARY` flag, and record this characteristic of link-layer device addresses in state.

## QA steps

- Bootstrap to AWS.
- Ensure we're logging in detail: `juju model-config logging-config="<root>=DEBUG"`.
- Check that `juju model-config container-networking-method` returns "fan".
- `juju add-machine --series bionic`.
- `juju add-machine lxd:0 --series bionic`.
- One quiesced, connect to Mongo and check that we are correctly acquiring link-layer devices and addresses.
- Specifically, see that we are detecting bridges and the parent/child relationships (excerpt from `db.linklayerdevices.find().pretty()`):
```
{
        "_id" : "d222ad4b-bdee-4542-86b9-8d54d0f61682:m#0#d#fan-252",
        "name" : "fan-252",
        "model-uuid" : "d222ad4b-bdee-4542-86b9-8d54d0f61682",
        "mtu" : 1450,
        "machine-id" : "0",
        "type" : "bridge",
        "mac-address" : "52:bf:50:8c:e6:2f",
        "is-auto-start" : true,
        "is-up" : true,
        "parent-name" : "",
        "virtual-port-type" : "",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "603f8f238f8c47104df4a487_ccef94e9"
        ]
}
{
        "_id" : "af900c3c-7e34-44be-8f81-bff80d8f7c14:m#0#d#ftun0",
        "name" : "ftun0",
        "model-uuid" : "af900c3c-7e34-44be-8f81-bff80d8f7c14",
        "mtu" : 8951,
        "machine-id" : "0",
        "type" : "ethernet",
        "mac-address" : "ea:2d:96:60:ac:a4",
        "is-auto-start" : true,
        "is-up" : true,
        "parent-name" : "fan-252",
        "virtual-port-type" : "",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "603f915e8f8c47104df4a738_7c9cd3a5"
        ]
}
```
- And `db.ip.addresses.find().pretty()` should have entries:
```
{
        "_id" : "af900c3c-7e34-44be-8f81-bff80d8f7c14:m#0/lxd/0#d#eth0#ip#252.18.31.111",
        "model-uuid" : "af900c3c-7e34-44be-8f81-bff80d8f7c14",
        "device-name" : "eth0",
        "machine-id" : "0/lxd/0",
        "subnet-cidr" : "252.16.0.0/12",
        "config-method" : "static",
        "value" : "252.18.31.111",
        "gateway-address" : "252.18.31.1",
        "is-default-gateway" : true,
        "origin" : "machine",
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "603f91a88f8c47104df4a86b_f509c60f"
        ]
}
```

## Documentation changes

None.

## Bug reference

Advances https://bugs.launchpad.net/juju/+bug/1863916
